nexus/operation-ids.md

```diff
@@ -0,0 +1,172 @@
+multi-tenant applications.
+
+In order to support stateless clients, the handler may provide an additional set of APIs to resolve client input to a
+reference Id. This may be something that is built into the protocol or implemented ad-hoc. To support this at the
+protocol level, we would need to add the following APIs:
+
+- `Resolve(operation, input) -> reference_id`
+
+-- OR --
+
+- `Cancel(operation, input)` - cancel operation by handler derived Id
+- `GetResult(operation, input)` - get operation result by handler derived Id
+- `Describe(operation, input)` - get operation status by handler derived Id
+
+NOTE: This looks a lot like the APIs for the "client generated Ids" approach, except the ID is structured instead of a
+single string.
+
+#### Cancel before started
+
+A caller may cancel an operation before it obtains a reference Id, e.g. if it is uncertain that the operation has been
+already started due to crashes and timeouts. Because the caller doesn't have a referencable Id or knowledge of which
+input fields are used to derive the Id, it would need to pass in the entire Start call input.
+
+This is especially important when we want to automate cancellations of workflow initiated operations, Temporal server
+cannot derive the Id from opaque input.
+
+### Comparing the approaches
+
+#### Handler Id qualification
+
+With handler generated Ids, to support stateless clients, the handler would need to be able to map an operation and
+input to a reference Id. It would also need to qualify a reference Id or securely issue the reference Ids when callers
+cannot be trusted. With caller generated Ids, the handler would only need to qualify an Id based on the caller context
+(e.g by concatinating it with a tenant Id).
+
+#### Ids as inputs
+
+With the caller generated Id approach, if the Id is used for handler side addressability, the handler would need to
+parse it and merge with the request's input.
+
+#### Caller friendliness
+
+Shifting the responsibility for creating a handler addressable Id to the caller makes an API harder to use.
+
+If we can map arbitrary client Ids to handler addressable resources, the client generated Id approach is easiest to use
+from stateless clients. The SDK can generate deterministic Ids for calls started from workflows.
+
+Handler generated Ids is the least friendly approach for stateless clients without additional APIs for mapping input to
+a handler addressable resources.
+
+#### Data sensitivity
+
+If a client is expected to provide a business meaningful Id that encodes sensitive information to address an operation,
+the Id would need to be encrypted making it harder to use.
+
+Both handler generated Ids and caller generated Ids with handler mapping options avoid the data sensitivity issue.
+
+## Proposal
+
+- Require client generated Ids for any call that starts an asynchronous operation.
+- API will be based on the rough draft in [Caller generated identifiers](#caller-generated-identifiers).
```

Comment:
I think this is a bad idea. It will make safe sharing much more difficult because the ids are not unguessable. And, in most cases, they cannot be completely client generated unless the handler leaks abstractions, or maintains bindings. 
I'm all for handler ids but  we need to help stateless clients maintain persistent aliases with "a service" (that may or not be this service).
However, we do need a client-created transaction id in Start() for "exactly-once", but that is not needed in other calls, and may not be part of the opaque identifier...

Reply:
Why can't IDs be guessable? We allow it with workflow IDs today and in general letting a record creator have the ability to set their own valuable-to-them primary key has value.

Reply:
I think the use case is a bit different. Sharing workflows IDs with others for collaboration is not that common today (I guess), with security an obvious problem, but when you go "higher-up-the-stack", people will start interacting with temporal in new ways, if we get the security right...
And I do agree with you, we need creators and consumers to make their own aliases to make this usable...